app/services/search_company_service.rb

```diff
@@ -0,0 +1,47 @@
+class SearchCompanyService
+
+  def initialize(params)
+    @params = params
+  end
+
+  def call
+    result = {
+      companies: [],
+      total_pages: 0,
+      errors: @errors
+    }
+    validator_result = Validators::CompanySearch.new(@params)
+    if validator_result.invalid?
+      result[:errors] = validator_result.errors.full_messages
+    else
+      search_company_result = search_company
+      result[:companies] = search_company_result[:companies]
+      result[:total_pages] = search_company_result[:total_pages]
+    end
+
+    result
+  end
+
+  private
+
+  def search_company
+    page = @params[:page] || 1
+    companies = Company.where(nil)
+    companies = companies.filter_by_industry(@params[:industry]) if @params[:industry].present?
+    companies = companies.filter_by_name(@params[:name]) if @params[:name].present?
+    companies = companies.filter_by_min_employee_count(@params[:min_employee_count]) if @params[:min_employee_count].present?
+    companies = companies.filter_by_min_deal_amount(@params[:minimum_deal_amount]) if @params[:minimum_deal_amount].present?
+    companies = companies.page(page).order(:name).fast_page
```

Comment:
fast_page is solving another known issue for paginating a large set, that each page is being more costly if we're using limit-offset, you can get more info about the problem & solution from [here](https://github.com/planetscale/fast_page?tab=readme-ov-file#how-this-works)